peer_gater.go

```diff
@@ -0,0 +1,252 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var PeerGaterRetainStats = 6 * time.Hour
+var PeerGaterQuiet = time.Minute
+
+type peerGater struct {
+	sync.Mutex
+
+	threshold, decay   float64
+	validate, throttle float64
+
+	lastThrottle time.Time
+
+	stats map[peer.ID]*peerGaterStats
+}
+
+type peerGaterStats struct {
+	connected bool
+	expire    time.Time
+
+	deliver, duplicate, ignore, reject float64
+}
+
+// WithPeerGater is a gossipsub router option that enables reactive validation queue
+// management.
+// The threshold parameter is the threshold of throttled/validated messages before the gating
+// kicks in.
+// The decay parameter is the (linear) decay of counters per second.
+func WithPeerGater(threshold, decay float64) Option {
+	return func(ps *PubSub) error {
+		gs, ok := ps.rt.(*GossipSubRouter)
+		if !ok {
+			return fmt.Errorf("pubsub router is not gossipsub")
+		}
+
+		gs.gate = newPeerGater(ps.ctx, threshold, decay)
+
+		// hook the tracer
+		if ps.tracer != nil {
+			ps.tracer.internal = append(ps.tracer.internal, gs.gate)
+		} else {
+			ps.tracer = &pubsubTracer{
+				internal: []internalTracer{gs.gate},
+				pid:      ps.host.ID(),
+				msgID:    ps.msgID,
+			}
+		}
+
+		return nil
+	}
+}
+
+func newPeerGater(ctx context.Context, threshold, decay float64) *peerGater {
+	pg := &peerGater{
+		threshold: threshold,
+		decay:     decay,
+		stats:     make(map[peer.ID]*peerGaterStats),
+	}
+	go pg.background(ctx)
+	return pg
+}
+
+func (pg *peerGater) background(ctx context.Context) {
+	tick := time.NewTicker(DefaultDecayInterval)
+
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-tick.C:
+			pg.decayStats()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (pg *peerGater) decayStats() {
```

Comment:
uber nit for future reference: pass the time in here from tick to avoid querying the time twice.